{"cells":[{"cell_type":"markdown","metadata":{"id":"jBq5TNCbg2oT"},"source":["# Week 4: Intro to Differential Expression Analysis ðŸ§¬\n","\n","From a biological standpoint, it's understood that differential expression analysis explores the differences in gene expression between two separate groups. This can be achieved through performing bulk RNA sequencing (bulk RNA-seq) on various sample of interest and corresponding controls. A review of the protocol for bulk RNA-seq, from one the original papers which used this method, can be found through the following link https://pubmed.ncbi.nlm.nih.gov/18516045/ and extended reading on RNA-seq data processing can be found at https://statomics.github.io/SGA/sequencing_intro.html.\n","\n","![workflow.jpg](https://drive.google.com/uc?export=view&id=1B1SnsccJahXAPf7uZf7lcpI8nCEwedEq)\n","\n","In this module, we will shift our view to the computational examination of bulk RNA-Seq data. We'll delve into utilizing Python-based tools for this analysis, understanding the function parameters, and deciphering the outcomes.\n","\n","By the end of this module, you will be equipped to accomplish the following tasks:\n","\n","- Apply necessary preprocessing steps to the tabular data\n","- Perform exploratory data analysis to look at data quality\n","- Run the differential expression analysis\n","- Interpret the results of the identified differentially expressed genes (DEGs)\n","\n","Next module, we will expand upon your findings and progress to visualizations and subsequent in-depth analysis, an essential phase in the data analysis workflow. Visualization aids in a more intuitive grasp of the data and can reveal trends or insights that raw data may not readily show. Engaging in further downstream analysis will enable you to formulate more definitive conclusions, evaluate hypotheses, or develop predictive models grounded in your preliminary results."]},{"cell_type":"markdown","source":["\n","##Introduction\n","\n","The complexity of gene expression data requires sophisticated statistical software to accurately identify DEGs. These tools must be able to handle large datasets, correct for multiple testing, and account for variability within and between sample groups. Traditionally, this sort of analysis would be done in R, not for any major reason other than that varying organizations have already developed infrastructure for R that may support differential expression analysis. However, slowly but surely the models and algorithms are being transcribed to Python.\n","\n","We will focus on a Python package called `pyDESeq2`, a Python implementation of the popular R package known as `DESeq2`. This external package will do most of the heavy lifting for us, we just have to supply it with the correctly formatted data. You can find the original paper that presented `DESeq2` with the following link https://pubmed.ncbi.nlm.nih.gov/25516281/\n","\n","### Dataset\n","\n","We will examine a dataset that contains gene expression data from the smooth muscle tissue of the airways in asthma patients. This dataset includes two groups: a treatment group that has received the asthma medication dexamethasone, and a control group of asthma patients who have not received any medication. The dataset comprises a total of eight samples, with four samples per group. For the purposes of this lesson, we have pre-processed the extensive raw RNA-Seq data files by aligning, verifying, and quantifying them. This preprocessing leaves us with a table of raw transcript counts. In this module, you will be exploring a specific file \"airway_rawcounts.csv\" which possess the **count matrix**, a table listing the raw transcript counts for each gene in each sample. Let's begin by loading in and viewing this count matrix.\n","\n","You can view and download the dataset from [airway_rawcounts.csv](https://drive.google.com/file/d/1YpPVt8tdOA4CCZhyJAPfoHKemL4RjXQM/view?usp=drive_link)"],"metadata":{"id":"S6O-A56Qz3g-"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"bE9CdYFag2oV"},"outputs":[],"source":["import pandas as pd\n","\n","counts = pd.read_csv(\"airway_rawcounts.csv\")\n","\n","counts"]},{"cell_type":"markdown","metadata":{"id":"6kvGA5tHg2oW"},"source":["**Q1. By using multiple samples, we are comparing the distributions of the transcript abundance for one gene between two groups. What sort of statistical test can you relate this to?**\n","\n","**You may find commonly-used statistical tests here: https://medium.com/@anushka.da3/types-of-statistical-tests-b8ceb90e13b3**\n","\n","**Write Answer Below**"]},{"cell_type":"markdown","metadata":{"id":"rfB8QOoUg2oW"},"source":["## Data Preprocessing\n","\n","Our `counts` DataFrame has the samples as the columns and the genes as the rows. In our analysis, it's essential to focus on data that can give us meaningful insights. This means we can exclude any gene that hasn't shown activity across all of our samples. To refine our `counts` DataFrame, we'll perform a simple but effective data cleaning step.\n","\n","Here's how we'll do it:\n","\n","1. **Assign the row names of the DataFrame to the first column**: We will change the rownames to be the first column for formatting puposes. There is no need for the row columns to be just arbitrary numbers specifying the gene's position in the DataFrame."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"OWG_ViWeg2oX"},"outputs":[],"source":["# Setting rownames to Geneid\n","counts = counts.set_index('ensgene')"]},{"cell_type":"markdown","metadata":{"id":"jDdPU9eEg2oX"},"source":["2. **Calculating the sum of counts per gene**: We'll sum up the counts across all columns for each gene. This is done to identify genes that show no expression at all in any of the samples. Here, the parameter, **axis** specifies how we want to sum our samples. Setting it to 1 means that we would like the sum of each row."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"4YR_B9K2g2oY"},"outputs":[],"source":["# Calculate the sum of counts for each gene across all samples\n","row_sums = counts.sum(axis=1)"]},{"cell_type":"markdown","metadata":{"id":"J4GU3h7hg2oY"},"source":["3. **Removing non-informative data our DataFrame**: We'll modify the `counts` DataFrame to keep only the rows (genes) that have a sum of counts greater than zero."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7M52TJivg2oZ"},"outputs":[],"source":["# Filter the DataFrame to retain only rows with nonzero total counts\n","counts = counts[row_sums > 0]"]},{"cell_type":"markdown","metadata":{"id":"6z421MVMg2oZ"},"source":["We can relate the samples as the observations, and the genes as the variables/features of our dataset. Thus, it would be conventional to swap the columns and rows so that the samples are the rows while the variables are the columns. Moreover, when passing this table to pyDESeq2, it is assumed that the samples are the rows."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"uv8iPI6vg2oZ"},"outputs":[],"source":["# Transpose counts table\n","counts = counts.T\n","\n","counts"]},{"cell_type":"markdown","metadata":{"id":"s2DXYpjng2oa"},"source":["To perform a detailed differential expression analysis, it's essential to consider the experiment's context, including each sample's collection conditions. This context is known as metadata, or \"sample information\" in DESeq2, referred to as `colData`.\n","\n","The count matrix holds the gene expression data but lacks the experimental condition descriptions for each sample, crucial for the analysis. To remedy this, we need to create a metadata DataFrame labeling each sample's condition as \"treatment\" (`T`) or \"control\" (`C`).\n","\n","We'll match each sample to its condition in the metadata DataFrame, ensuring it aligns with the count matrix's sample order. This involves making a concise pandas DataFrame with one column, where labels correspond to the count matrix's sample order, facilitating accurate data analysis."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"hP3TbSaDg2oa","outputId":"0f272b9e-cb9f-48d3-f46e-a55e76642023","colab":{"base_uri":"https://localhost:8080/","height":300},"executionInfo":{"status":"ok","timestamp":1722836295447,"user_tz":240,"elapsed":12,"user":{"displayName":"Yifei Sun","userId":"08728652971139677065"}}},"outputs":[{"output_type":"execute_result","data":{"text/plain":["           Condition\n","SRR1039508         C\n","SRR1039509         T\n","SRR1039512         C\n","SRR1039513         T\n","SRR1039516         C\n","SRR1039517         T\n","SRR1039520         C\n","SRR1039521         T"],"text/html":["\n","  <div id=\"df-6fe46a2c-2e2c-4d29-ab30-342b0f83c75c\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Condition</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>SRR1039508</th>\n","      <td>C</td>\n","    </tr>\n","    <tr>\n","      <th>SRR1039509</th>\n","      <td>T</td>\n","    </tr>\n","    <tr>\n","      <th>SRR1039512</th>\n","      <td>C</td>\n","    </tr>\n","    <tr>\n","      <th>SRR1039513</th>\n","      <td>T</td>\n","    </tr>\n","    <tr>\n","      <th>SRR1039516</th>\n","      <td>C</td>\n","    </tr>\n","    <tr>\n","      <th>SRR1039517</th>\n","      <td>T</td>\n","    </tr>\n","    <tr>\n","      <th>SRR1039520</th>\n","      <td>C</td>\n","    </tr>\n","    <tr>\n","      <th>SRR1039521</th>\n","      <td>T</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6fe46a2c-2e2c-4d29-ab30-342b0f83c75c')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-6fe46a2c-2e2c-4d29-ab30-342b0f83c75c button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-6fe46a2c-2e2c-4d29-ab30-342b0f83c75c');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-ea9ad4e5-bb78-4425-9884-444064e74918\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-ea9ad4e5-bb78-4425-9884-444064e74918')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-ea9ad4e5-bb78-4425-9884-444064e74918 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","  <div id=\"id_3aecfd5b-6ef7-4af0-9a5f-5efc3430a4fa\">\n","    <style>\n","      .colab-df-generate {\n","        background-color: #E8F0FE;\n","        border: none;\n","        border-radius: 50%;\n","        cursor: pointer;\n","        display: none;\n","        fill: #1967D2;\n","        height: 32px;\n","        padding: 0 0 0 0;\n","        width: 32px;\n","      }\n","\n","      .colab-df-generate:hover {\n","        background-color: #E2EBFA;\n","        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","        fill: #174EA6;\n","      }\n","\n","      [theme=dark] .colab-df-generate {\n","        background-color: #3B4455;\n","        fill: #D2E3FC;\n","      }\n","\n","      [theme=dark] .colab-df-generate:hover {\n","        background-color: #434B5C;\n","        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","        fill: #FFFFFF;\n","      }\n","    </style>\n","    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('metadata')\"\n","            title=\"Generate code using this dataframe.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n","  </svg>\n","    </button>\n","    <script>\n","      (() => {\n","      const buttonEl =\n","        document.querySelector('#id_3aecfd5b-6ef7-4af0-9a5f-5efc3430a4fa button.colab-df-generate');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      buttonEl.onclick = () => {\n","        google.colab.notebook.generateWithVariable('metadata');\n","      }\n","      })();\n","    </script>\n","  </div>\n","\n","    </div>\n","  </div>\n"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"dataframe","variable_name":"metadata","summary":"{\n  \"name\": \"metadata\",\n  \"rows\": 8,\n  \"fields\": [\n    {\n      \"column\": \"Condition\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"T\",\n          \"C\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"}},"metadata":{},"execution_count":6}],"source":["# Create a DataFrame with sample conditions, using the index from the count matrix\n","metadata = pd.DataFrame({\n","    'Condition': ['C', 'T', 'C', 'T', 'C', 'T', 'C', 'T']}, index=counts.index)\n","metadata"]},{"cell_type":"markdown","metadata":{"id":"mp7fQF-0g2oa"},"source":["Let us see what happens when we do not set up the metadata of the count matrix correctly.\n","\n","Run the code below. You may notice that there is an extra `T` at the end of the list.\n","\n","The error involves the fact that there is a mismatch between the `Condition` list and the number of samples in the count matrix. The line of code knows how many samples there are in the count matrix for it is included when we specified the index parameter."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"T9BGeSQ-g2ob"},"outputs":[],"source":["# Create a DataFrame with sample conditions, using the index from the count matrix\n","test_metadata = pd.DataFrame({\n","    'Condition': ['C', 'T', 'C', 'T', 'C', 'T', 'C', 'T', 'T']}, index = counts.index)\n","test_metadata"]},{"cell_type":"markdown","metadata":{"id":"0BOeQXD1g2ob"},"source":["## Exploratory Data Analysis\n","\n","In this section, we learn how we do hypothesis testing and differential gene expression analysis of RNA-seq data using pyDESeq\n","\n","Now we have set up our data to be used by pyDESeq, but before we run our extensive analysis, it's first a good idea to get some preliminary analytics as to how well our groups may be distinguished from each other. There are many ways to do this, but one common was is to use principle component analysis (PCA), which you have learned in the last module.\n","\n","PCA is a common pick because we are working with data utilizing genes as feature; thus, our data is of high dimensionality. Moreover, We would like to be able to see a visualization as to how well the data clusters into the two groups; thus, we need to reduce the number of features to 2 (or 3 as a 3D plot is possible as well).\n","\n","The PCA results helps us visualize the clustering of samples based on their gene expression profiles. This is useful in identifying groups of samples that are similar to each other, which might indicate similar biological conditions, treatment groups, or technical batches.\n","\n","Below is the implementation of a PCA plot labelled using the condition of the sample. We are hoping to see that the two groups cluster well and there is minimal overlap."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"MgiCGGnzg2ob"},"outputs":[],"source":["from sklearn.decomposition import PCA\n","from sklearn.preprocessing import StandardScaler\n","import matplotlib.pyplot as plt\n","\n","# Step 1: Standardize the count matrix\n","scaler = StandardScaler()\n","scaled_counts = scaler.fit_transform(counts)\n","\n","# Step 2: Compute PCA\n","pca = PCA(n_components=2)  # Choosing 2 components for a 2D plot\n","pca_result = pca.fit_transform(scaled_counts)\n","\n","# Step 3: Prepare the data for plotting\n","# Convert the PCA results into a DataFrame\n","pca_df = pd.DataFrame(data=pca_result, columns=['PC1', 'PC2'], index=counts.index)\n","\n","# Join the PCA results with the metadata to align the samples with their labels\n","plot_data = pca_df.join(metadata)\n","\n","# Step 4: Plot the PCA\n","plt.figure(figsize=(5, 5))\n","for condition in plot_data['Condition'].unique():\n","    # Select data points belonging to the current condition\n","    subset = plot_data[plot_data['Condition'] == condition]\n","    plt.scatter(subset['PC1'], subset['PC2'], label=condition, alpha=0.7, edgecolors='w')\n","\n","plt.title('PCA of Count Matrix')\n","plt.xlabel('Principal Component 1')\n","plt.ylabel('Principal Component 2')\n","plt.legend()\n","plt.show()\n"]},{"cell_type":"markdown","source":["**Q2: Does Principal Component 1 or Principal Component 2 explain the separation between the data points of treatment and control groups better? Explain your reasoning.**\n","\n","**Write Answer Below**"],"metadata":{"id":"Ka2vuv4EHpTF"}},{"cell_type":"markdown","metadata":{"id":"B173hioqg2ob"},"source":["## Differential Expression Analysis\n","\n","To practically implement differential expression analysis using PyDESeq2, we begin by constructing a DESeqDataSet object, commonly referred to as `dds` in code examples. This object is central to the PyDESeq2 workflow as it encapsulates all the data required for the analysis, including the count data and the associated sample metadata.\n","\n","The creation of the `dds` object is not just a formality but a critical step that ensures data integrity and prepares the dataset for the subsequent analysis stages. It checks for consistency between the count data and the metadata, and it sets the stage for incorporating the experimental design into the analysis.\n","\n","Here's how we create the DESeqDataSet object in practice:\n","\n","1. **Prepare the count data and metadata**: Ensure that your count matrix and metadata DataFrame are correctly formatted and aligned.\n","\n","2. **Define the experimental design**: Specify the design of your experiment by indicating which column in the metadata DataFrame represents the condition or factor of interest.\n","\n","3. **Create the DESeqDataSet object**: Use the count data and metadata to initialize the DESeqDataSet object, incorporating the experimental design.\n","\n","The Python code to create the \"dds\" object looks like what's below, but enter the corresponding variables for the count matrix and metadata table that we have prepared:\n","\n","**Q3: We left out the parameter values for you. Please adjust the code to correctly use the input data as parameters.**"]},{"cell_type":"code","source":["!pip install pydeseq2"],"metadata":{"id":"Tc-B5yC1GAe_"},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"id":"uoz5ouC0g2ob"},"outputs":[],"source":["from pydeseq2.dds import DeseqDataSet\n","from pydeseq2.ds import DeseqStats\n","\n","# TODO: Enter counts and metadata parameters\n","dds = DeseqDataSet(counts = , metadata = , design_factors = 'Condition')"]},{"cell_type":"markdown","metadata":{"id":"CupJcFLmg2oc"},"source":["Below we will finally run the differential expression analysis and compare the distributions of each gene. To acheive this, we simply call the `deseq2` function. It carries out several steps to determine which genes or transcripts are differentially expressed across your conditions of interest:\n","\n","- **Normalization**: DESeq2 employs data adjustment techniques to ensure a fair comparison of gene expression between samples, similar to ensuring that two fruit baskets being compared are of the same size. This normalization process in DESeq2 is performed to remove biases stemming from differences in sequencing depths or other technical factors, making it possible to accurately assess gene expression changes across samples.\n","\n","- **Variability Estimation**: It calculates how much the expression of each gene varies naturally, so we can distinguish real differences from random noise. This is based on the variability observed across all samples within the experimental conditions or groups being compared. In fact, DESeq2 requires at least 2 replicates per group for the analysis (as a bare minimum; at least 3 however, is usually recommended).\n","\n","- **Testing for Differences**: DESeq2 then identifies genes that are consistently expressed differently between conditions, beyond what would be expected by chance.\n","\n","- **Results**: It provides a list of these genes, along with the strength and certainty of their differential expression."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"6_HOMS01g2oc"},"outputs":[],"source":["dds.deseq2()"]},{"cell_type":"markdown","metadata":{"id":"ritmGTo9g2oc"},"source":["After you run the analysis with DESeq2, your data object gets updated with new information. This includes adjustments made to account for different amounts of data across samples, measurements of variability for each gene, and the estimated changes in gene expression. While the detailed results aren't stored directly in this object, it does hold key figures and notes about the analysis that can be useful for further examination and interpretation.\n","\n","To obtain the results of the differential expression tests we must construct a `DeseqStats` object using our `DESeqDataSet` object. In this object, we must specify which 2 groups we are comparing based on our metadata. As you can expect, we are comparing based on \"Condition\", the column name in the metadata table. We are specifically comparing between `T`, samples treated with dexmethasone, and `C`, control samples. Finally, by calling summary, it runs the statistical analysis and generates us a comprehensive table with our statistical results.\n","\n","From these statistical results, you may already be able to draw biological interpretations and conclusions, Thus, it would be most beneficial to get familiar with what each of these columns represent:\n","\n","| Term             | Description |\n","|------------------|-------------|\n","| `baseMean`       | The average of the normalized counts for a gene across all samples and conditions. |\n","| `log2FoldChange` | The log (base 2) of the fold change between conditions. Positive values indicate up-regulation, and negative values indicate down-regulation. |\n","| `lfcSE`          | The standard error of the log2 fold change, indicating the uncertainty of the `log2FoldChange` estimate. |\n","| `stat`           | The Wald test statistic used to test the significance of the `log2FoldChange`. |\n","| `pvalue`         | The p-value from the Wald test, showing the probability of observing the data if there was no true difference in expression. |\n","| `padj`           | The adjusted p-value accounting for multiple testing, reducing the chance of false positives in significant gene identification. |\n","\n","The `pvalue` measures the chance of observing gene expression differences if no actual difference exists, identifying significant changes. However, with multiple tests, false positives increase. The `padj` corrects for this by adjusting p-values for multiple comparisons, using methods like Bonferroni, Benjamini-Hochberg, and FDR to control the overall error rate, typically set at 0.05.\n","\n","Fold changes show gene expression differences between conditions, with their significance depending on the analysis order. For \"Condition T vs C,\" a positive fold change means treatment (T) has more transcripts than control (C), emphasizing the importance of condition order in result interpretation."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bUwzUelFg2oc"},"outputs":[],"source":["# Calculate the statistical results\n","res = DeseqStats(dds, contrast=('Condition', 'T', 'C'))\n","res.summary()"]},{"cell_type":"markdown","metadata":{"id":"Losiqrz_g2oc"},"source":["Now, we would like this table above as specifically a pandas dataframe, so we could make adjustments to it such as futher filtering. We will save this new dataframe as `res_df`."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"5Iy87cTcg2od","outputId":"1c12e3f4-149d-4305-e0ae-b5643f4febd3","colab":{"base_uri":"https://localhost:8080/","height":455},"executionInfo":{"status":"ok","timestamp":1722836531572,"user_tz":240,"elapsed":221,"user":{"displayName":"Yifei Sun","userId":"08728652971139677065"}}},"outputs":[{"output_type":"execute_result","data":{"text/plain":["                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n","ensgene                                                                     \n","ENSG00000000003  708.602170       -0.378860  0.173829 -2.179501  0.029294   \n","ENSG00000000419  520.297901        0.203748  0.099570  2.046276  0.040729   \n","ENSG00000000457  237.163037        0.034101  0.124390  0.274146  0.783973   \n","ENSG00000000460   57.932633       -0.117126  0.296504 -0.395022  0.692826   \n","ENSG00000000938    0.318098       -1.726440  3.586871 -0.481322  0.630288   \n","...                     ...             ...       ...       ...       ...   \n","ENSG00000273487    8.163235        1.014039  0.556511  1.822137  0.068434   \n","ENSG00000273488    8.584479        0.221732  0.550367  0.402880  0.687037   \n","ENSG00000273489    0.275899        1.546042  3.724653  0.415084  0.678081   \n","ENSG00000273492    0.105978       -0.540731  3.871056 -0.139686  0.888908   \n","ENSG00000273493    0.106142       -0.540731  3.871056 -0.139686  0.888908   \n","\n","                     padj  \n","ensgene                    \n","ENSG00000000003  0.143413  \n","ENSG00000000419  0.178607  \n","ENSG00000000457  0.929945  \n","ENSG00000000460  0.893503  \n","ENSG00000000938       NaN  \n","...                   ...  \n","ENSG00000273487  0.254193  \n","ENSG00000273488  0.890211  \n","ENSG00000273489       NaN  \n","ENSG00000273492       NaN  \n","ENSG00000273493       NaN  \n","\n","[33469 rows x 6 columns]"],"text/html":["\n","  <div id=\"df-2a6c3eb5-537e-4ecf-94f8-38acafb03371\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>baseMean</th>\n","      <th>log2FoldChange</th>\n","      <th>lfcSE</th>\n","      <th>stat</th>\n","      <th>pvalue</th>\n","      <th>padj</th>\n","    </tr>\n","    <tr>\n","      <th>ensgene</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>ENSG00000000003</th>\n","      <td>708.602170</td>\n","      <td>-0.378860</td>\n","      <td>0.173829</td>\n","      <td>-2.179501</td>\n","      <td>0.029294</td>\n","      <td>0.143413</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000000419</th>\n","      <td>520.297901</td>\n","      <td>0.203748</td>\n","      <td>0.099570</td>\n","      <td>2.046276</td>\n","      <td>0.040729</td>\n","      <td>0.178607</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000000457</th>\n","      <td>237.163037</td>\n","      <td>0.034101</td>\n","      <td>0.124390</td>\n","      <td>0.274146</td>\n","      <td>0.783973</td>\n","      <td>0.929945</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000000460</th>\n","      <td>57.932633</td>\n","      <td>-0.117126</td>\n","      <td>0.296504</td>\n","      <td>-0.395022</td>\n","      <td>0.692826</td>\n","      <td>0.893503</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000000938</th>\n","      <td>0.318098</td>\n","      <td>-1.726440</td>\n","      <td>3.586871</td>\n","      <td>-0.481322</td>\n","      <td>0.630288</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000273487</th>\n","      <td>8.163235</td>\n","      <td>1.014039</td>\n","      <td>0.556511</td>\n","      <td>1.822137</td>\n","      <td>0.068434</td>\n","      <td>0.254193</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000273488</th>\n","      <td>8.584479</td>\n","      <td>0.221732</td>\n","      <td>0.550367</td>\n","      <td>0.402880</td>\n","      <td>0.687037</td>\n","      <td>0.890211</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000273489</th>\n","      <td>0.275899</td>\n","      <td>1.546042</td>\n","      <td>3.724653</td>\n","      <td>0.415084</td>\n","      <td>0.678081</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000273492</th>\n","      <td>0.105978</td>\n","      <td>-0.540731</td>\n","      <td>3.871056</td>\n","      <td>-0.139686</td>\n","      <td>0.888908</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>ENSG00000273493</th>\n","      <td>0.106142</td>\n","      <td>-0.540731</td>\n","      <td>3.871056</td>\n","      <td>-0.139686</td>\n","      <td>0.888908</td>\n","      <td>NaN</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>33469 rows Ã— 6 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2a6c3eb5-537e-4ecf-94f8-38acafb03371')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-2a6c3eb5-537e-4ecf-94f8-38acafb03371 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-2a6c3eb5-537e-4ecf-94f8-38acafb03371');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-bf6abc71-e300-42dd-a869-29ce9ca32005\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-bf6abc71-e300-42dd-a869-29ce9ca32005')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-bf6abc71-e300-42dd-a869-29ce9ca32005 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","  <div id=\"id_b4a45785-ff96-4099-ad50-fd5fd20a42a3\">\n","    <style>\n","      .colab-df-generate {\n","        background-color: #E8F0FE;\n","        border: none;\n","        border-radius: 50%;\n","        cursor: pointer;\n","        display: none;\n","        fill: #1967D2;\n","        height: 32px;\n","        padding: 0 0 0 0;\n","        width: 32px;\n","      }\n","\n","      .colab-df-generate:hover {\n","        background-color: #E2EBFA;\n","        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","        fill: #174EA6;\n","      }\n","\n","      [theme=dark] .colab-df-generate {\n","        background-color: #3B4455;\n","        fill: #D2E3FC;\n","      }\n","\n","      [theme=dark] .colab-df-generate:hover {\n","        background-color: #434B5C;\n","        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","        fill: #FFFFFF;\n","      }\n","    </style>\n","    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('res_df')\"\n","            title=\"Generate code using this dataframe.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n","  </svg>\n","    </button>\n","    <script>\n","      (() => {\n","      const buttonEl =\n","        document.querySelector('#id_b4a45785-ff96-4099-ad50-fd5fd20a42a3 button.colab-df-generate');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      buttonEl.onclick = () => {\n","        google.colab.notebook.generateWithVariable('res_df');\n","      }\n","      })();\n","    </script>\n","  </div>\n","\n","    </div>\n","  </div>\n"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"dataframe","variable_name":"res_df","summary":"{\n  \"name\": \"res_df\",\n  \"rows\": 33469,\n  \"fields\": [\n    {\n      \"column\": \"ensgene\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 33469,\n        \"samples\": [\n          \"ENSG00000232023\",\n          \"ENSG00000175084\",\n          \"ENSG00000151963\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"baseMean\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 5015.410103510234,\n        \"min\": 0.08934720627807478,\n        \"max\": 325662.6556601783,\n        \"num_unique_values\": 24897,\n        \"samples\": [\n          158.12959870107917,\n          260.0225705402865,\n          371.1446931631407\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"log2FoldChange\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.8825902456301888,\n        \"min\": -5.405196983693194,\n        \"max\": 9.609711013470308,\n        \"num_unique_values\": 24882,\n        \"samples\": [\n          -0.28266347787573637,\n          -1.0829394478901178,\n          -0.10444263944361708\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"lfcSE\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 1.4076785353595032,\n        \"min\": 0.058815462600605314,\n        \"max\": 3.871082102113682,\n        \"num_unique_values\": 24881,\n        \"samples\": [\n          0.19020727654546232,\n          0.35403424882977574,\n          0.8830858428821154\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"stat\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 1.5963550726616877,\n        \"min\": -13.853383499734942,\n        \"max\": 21.589328965707953,\n        \"num_unique_values\": 24882,\n        \"samples\": [\n          -0.4898808410662373,\n          -0.44683651439862715,\n          -0.8303952224441132\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pvalue\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.33015173053164387,\n        \"min\": 2.262742083130443e-103,\n        \"max\": 0.999987641385527,\n        \"num_unique_values\": 24831,\n        \"samples\": [\n          0.03507924307960741,\n          0.20796127704984568,\n          0.42026138500455856\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"padj\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.3444075152184504,\n        \"min\": 4.331340895528294e-99,\n        \"max\": 0.999987641385527,\n        \"num_unique_values\": 10235,\n        \"samples\": [\n          1.7264638393419151e-06,\n          0.011695475159139065,\n          2.3071695801483047e-05\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"}},"metadata":{},"execution_count":13}],"source":["# Extracting the results dataframe\n","res_df = res.results_df\n","res_df"]},{"cell_type":"markdown","metadata":{"id":"1XckXhn_g2od"},"source":["You may have noticed that these genes are labeled with their **Ensembl IDs**. Although this naming scheme is very structured and orderly, it is very hard for an individual to decipher and identify genes of interest, akin to trying to identify a product at a grocery store by reading its barcode instead of reading the label. Therefore, we will map these Ensembl IDs to their Gene Symbol counterparts through the use of a dictionary obtained from an external library that we import. In the results dataframe, we will add a new column that consists of these gene names."]},{"cell_type":"code","source":["!pip install sanbomics"],"metadata":{"id":"yShAK4AIIHXi"},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"id":"WSSwsvOXg2od"},"outputs":[],"source":["# Package to help perform downstream analysis\n","from sanbomics.tools import id_map\n","\n","# Obtaining the dictionary mapping ensembl ids to Gene symbols\n","human_mapper = id_map(species = 'human')\n","\n","# Adding a column of the gene names.\n","res_df['Symbol'] = res_df.index.map(human_mapper.mapper)\n","res_df"]},{"cell_type":"markdown","metadata":{"id":"sz2fgW45g2od"},"source":["Using this dataframe, we aim to pinpoint genes that are significantly affected. Not all genes with detected expression will be significant. To find the significant ones, we'll look for genes with an adjusted p-value less than 0.05 and an absolute log2 fold change value greater than 0.5 (or approx. 1.41x). We'll also exclude genes with a baseMean value under 10 to avoid genes that are expressed at very low levels, as these can show misleading variability. The cutoff of 10 is a commonly used benchmark, but like the p-value and fold-change thresholds, this value is arbitrarily selected."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"WEPIUkdjg2od"},"outputs":[],"source":["# Filtering out genes that have a low amout of reads\n","read_thresh = 10\n","res_df = res_df[res_df.baseMean >= read_thresh]\n","# Filter the results dataframe 'res_df' for significant genes\n","# Start by identifying genes with an adjusted p-value less than 0.05\n","significant_p_value = res_df.padj < 0.05\n","# Next, identify genes with an absolute log2 fold change greater than 1\n","significant_fold_change = abs(res_df.log2FoldChange) > 1\n","# Combine both conditions to get genes that satisfy both criteria\n","sigs = res_df[significant_p_value & significant_fold_change]\n","# Display the top 15 differentially expressed genes from the filtered list\n","sigs.sort_values(by=['log2FoldChange'], ascending=False).head(15)"]},{"cell_type":"markdown","metadata":{"id":"NMfspBF6g2oe"},"source":["**Q4: Examine the LogFoldChanges of the genes. Keep in mind that we aim to identify genes that are either upregulated or downregulated. In the subsequent cell, please answer the question as to why a gene with a log fold change close to 0 might still exhibit a significant p-value.**\n","\n","**Write Answer Below**"]},{"cell_type":"markdown","source":["**Q5: Further examine the list of the top 15 differentially expressed genes by searching the ensemble gene code, given to you in the ensgene column, for each of the top 15 genes in GeneCards, a database of annotated human genes with genomic, transcriptomic, proteomic, genetic, clinical and functional information (https://www.genecards.org/). List 2-3 genes from the top 15 list of genes that share a common function and explain why this function/pathway may be upregulated in the treatment group.**\n","\n","(Hint: the genes with the following ensemble gene codes may have some cellular pathways in common, ENSG00000109906 ENSG00000171819 ENSG00000168309 ENSG00000152583 ENSG00000127324)\n","\n","**Write Your Answer Below**\n"],"metadata":{"id":"-Xn4NjSJXczX"}},{"cell_type":"markdown","metadata":{"id":"8_bw0ETgg2oe"},"source":["## Saving our Results\n","\n","Next week, we will be building off the results we obtained from this week. Below is the code to save your file if you are curious in continuing the analysis with your results. Otherwise, a results file will be provided for you next week. We will be saving both the complete results dataframe."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"u__H1ISig2oe"},"outputs":[],"source":["# Save the DataFrame as a CSV file\n","res_df.to_csv('DGE_complete_results.csv', index=True)"]},{"cell_type":"markdown","metadata":{"id":"6TVROcH6g2og"},"source":["## Conclusion\n","\n","In this module, you learned how to preprocess a tabular data file containing the raw counts of a bulk RNA-seq experiment. We then performed differntial gene expression analysis using a Python implementation of DESeq2. Finally we obtained our results and aimed to interpret them both statistically and biologically. As well, we have saved the dataframe containing the results for later use next module."]},{"cell_type":"markdown","metadata":{"id":"xYsa6dSQg2oe"},"source":["## Graded Exercises (5 marks)"]},{"cell_type":"markdown","metadata":{"id":"LZWV4oqXg2oe"},"source":["**GQ1: We used the following line to instatiate the object that will be holding our count data and metadata: `DeseqDataSet(counts = counts, metadata = metadata, design_factors ='Condition')`. Here is the documentation on the `DeseqDataSet` object: https://pydeseq2.readthedocs.io/en/latest/api/docstrings/pydeseq2.dds.DeseqDataSet.html. Explain what the parameter `design_factors` does. Also, what changes would you make to this line of code if you were instead interested in the difference in gene expression between a different variable? For example, if the sexes of the samples were disclosed in the metadata? (2 marks)**"]},{"cell_type":"markdown","metadata":{"id":"wIzYPL8Tg2of"},"source":["**GQ2: In their 2012 research, Menke and colleagues discovered that dexamethasone, a drug applied to our treatment samples, consistently regulates two genes: FKBP5 and DUSP1. Suppose we aim to corroborate this finding that the genes are regulated. Given the dataframe of significant results, extract the data pertaining to FKBP5 and DUSP1.**\n","\n","**Hint: remember to look up the names using the `Symbol` column, not the `ensgene`. (1 mark)**\n","\n","https://www.nature.com/articles/npp2011331"]},{"cell_type":"markdown","metadata":{"id":"DeD-EIg7g2of"},"source":["**GQ3: Based on your finding in the previous question, answer the following questions: (2 marks)**\n","- Are the results significant and are the genes upregulated or downregulated?\n","- Are the claims valid that the medication changes the regulation of these genes?\n","- Can you look up and find any information as to what these two genes are reponsible for? Feel free to read the paper that was linked as well."]}],"metadata":{"kernelspec":{"display_name":"tensor","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.8"},"colab":{"provenance":[{"file_id":"1drUx-C5oYL7Y9MfDcv0L58SuNcX9xAbg","timestamp":1724162890774},{"file_id":"183JUfF1k9dHVomnwnx6GgA1QW__RNlz0","timestamp":1722836117471}]}},"nbformat":4,"nbformat_minor":0}